// @rettersoft/retter-log/index.ts
import AWS from "aws-sdk";
import https from "https";

const {RETTERLOG_KINESIS} = process.env;

export default class RetterLog{
    private static readonly kinesis = new AWS.Kinesis({
        maxRetries: 6,
        httpOptions: {
            agent: new https.Agent({
                maxSockets: Infinity,
                keepAlive: true
            }),
            timeout: 5000,
            connectTimeout: 5000
        }
    });

    /**
     * Logs the request to the stream.
     *
     * @param request - The request that is going to be logged
     * @param response - The response of the request that is going to be logged
     * @param tag - The type of the request
     * @param host - The host of the log
     * @param [endpoint] - The endpoint of the log 
     * @returns The given response
     */
    public static async LogRequest(request: any, response: any, tag: string, host: string, endpoint?:string){
        try{
            if(!RETTERLOG_KINESIS){
                console.log("Please specify RETTERLOG_KINESIS as environment variable");
            }
            else{
                await RetterLog.kinesis.putRecord({
                    StreamName: RETTERLOG_KINESIS,
                    PartitionKey: Date.now().toString(),
                    Data: JSON.stringify({
                            request,
                            response,
                            tag,
                            host,
                            endpoint
                        })
                }).promise();
            }
            return response;      
        }
        catch(e){
            console.warn(e);
            return response;
        }
    }

}
